super-ls

e1528532 opened the pull request:
# Purpose

This branch is used to develop an advantaged version of the ls tool to include further information e.g. wheter if a key is a node or a leaf, and the number of nodes/leaves inside a given node as request in #1167 .

# Checklist

Please only check relevant points.
For docu fixes, spell checking and similar nothing
needs to be checked.

- [ ] commit messages are fine (with references to issues)
- [ ] I ran all tests and everything went fine
- [ ] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

@markus2330 Please note this is heavy WIP currently. You don't seem to like the name super-ls, but i think it suits fine as a working title for now until we find a proper name. Another idea i had is that we could just add an option to the original ls tool to display node/leaf/count information instead of handling this as a new tool. In that case we could also add an option to switch between newline or NULL separators. Or lets keep it a new tool?

And consider the following case of keys:
user/test/key = "test1"
user/test/key/2 = "test2"

should the output be like
user/test node 1
user/test/key node 1
user/test/key/2 leaf

or simply (so not listing the user/test node which does not actually exist in elektra in the example)
user/test/key node 1
user/test/key/2 leaf

the latter would be easier i think as it allows me to simply iterate over the keyset i get via kdbGet like i'm doing it now.
Looking forward to some feedback as its the first actual c++ code i'm doing for elektra

markus2330 commented:
> You don't seem to like the name super-ls

Yes, you are right ;) Maybe `completion-helper`? We should take it out from the list of commands, maybe even develop as external command?

>  Another idea i had is that we could just add an option to the original ls tool to display node/leaf/count information instead of handling this as a new tool.

Only if the feature is useful for humans, too. And this might be not the case, especially if we add the feature to complete to mountpoints.

We should dedicate the tool to the goal: "given a key name, give me possibilities how to complete the key name". This is something quite different to what `kdb ls` does.

> In that case we could also add an option to switch between newline or NULL separators.

`-0` already does that.

@reox

reox commented:
> We should dedicate the tool to the goal: "given a key name, give me possibilities how to complete the key name". This is something quite different to what kdb ls does.
Yes, that would be exactly what the completion file does. I get the current cmdline like `/sw/pk/`, then i query for all keys that are there and get for example `/sw/pk/foo/sdf/`, `/sw/pk/foo/sdfsdg`, `/sw/pk/bla/sdf` and  then i strip all the compontents from the front and back to have only the next level.

Maybe have it like `find` with --maxdepth and --mindepth to look up keys:
`super-ls --maxdepth=1 /sw/pk` would return `/sw/pk/foo` and `/sw/pk/bla/` (+all the other information like leaf and node)

e1528532 commented:
ok, i have created a prototype (yeah still called super-ls for now, will migrate it to an own tool or rename it to completion-helper once the functionality is more or less done). There is plenty of space for refactoring and optimizations but for now we should get the functionality straight before spending time on that.

As its still based on the kdb tool right now, i just reused the -r parameter (recursive) to display all possible completions, and otherwise its just the next level, so like --maxdepth=1 would be. Internally there is already a support for having a max and mindepth.

I think the easiest way to clarify the requirements is to show some examples how it currently works, based on this test dataset:
```
mbp-armin:build admin$ kdb ls user/
user/kdb_test/1/key_name
user/kdb_test/1/key_name/2
user/kdb_test/1/other
user/test
user/test/bla
user/test/blu
```

```
mbp-armin:build admin$ kdb super-ls user/
user/kdb_test node 1
user/test node 2
```

```
mbp-armin:build admin$ kdb super-ls user/kd
user/kdb_test node 1
```

```
mbp-armin:build admin$ kdb super-ls user/kdb_test
user/kdb_test/1 node 2
```

```
mbp-armin:build admin$ kdb super-ls -r user/kd
user/kdb_test node 1
user/kdb_test/1 node 2
user/kdb_test/1/key_name node 1
user/kdb_test/1/key_name/2 leaf
user/kdb_test/1/other leaf
```

For cascading keys, i return the absolute keynames so you clearly see to which key it would resolve which I think is the more helpful approach:
```
mbp-armin:build admin$ kdb super-ls -r /kd
user/kdb_test node 1
user/kdb_test/1 node 2
user/kdb_test/1/key_name node 1
user/kdb_test/1/key_name/2 leaf
user/kdb_test/1/other leaf
```

```
mbp-armin:build admin$ kdb super-ls /
system/elektra node 2
user/kdb_test node 1
user/test node 2
```

@reox is this going in the right direction?

reox commented:
Yes, i think this looks good!

e1528532 commented:
@markus2330 ok since it looks like i've hit the main requirements successfully, now its time to decide wheter it should be an external command or not.

I hope its ok that its done in c++ like the other kdb tools are, as i think i would have been much slower coding this in plain c. It still took me quite some time as i had to get familiar with c++ again and exploring how the elektra api can be used for a real world problem, handling cascading keys and "non existent keys" was also not too easy (with non existant keys i mean if there is a key "user/this/is/a/key", and i want to complete "user/this/is" which does not correspond to an actual existing key, and "user/this/is/a" being no existent key either, that it still suggests "user/this/is/a" which would be a node then with the child key).

Leaving it as a kdb tool would have the benefit that its not much more work, but if all the kdb tools are intended for humans, its probably not the best idea.

If we want to make an own tool out of it, how would it be called, kdb-completion-helper? kdb-complete? kdb-suggest? I guess i could reuse most of the command line parsing from kdb, and i think thats all i need apart from my super-ls files.

Ah and a last thing, I yet have to do the mountpoint completion. Do i get it right that i can fetch a current list of mountpoints from system/elektra/mountpoints which I'd parse and add to the suggestion processing then? If so, i think its not too difficult to implement it. Should it somehow be marked as a mountpoint in the output? Its very nice to see how elektra uses itself to store parts of its configuration, clearly demonstrates its capabilities!

markus2330 commented:
> Leaving it as a kdb tool would have the benefit that its not much more work

 :+1:  Sounds like a good idea then!

> but if all the kdb tools are intended for humans, its probably not the best idea.

We already have this issue with `list-commands` which is not really useful for users, too. (Any mistyped command gives you the list.) So I would suggest we add a feature in the factory to hide internal/renamed commands or even aliases (gmount vs. global-mount) See #265 to have an idea about the long-term goal.

Then we would simply hide it from the list.

markus2330 commented:
Implementation note: better to not mess around with the factory, but directly improve the `Command` or `Instancer` class. `getLongHelpText` actually is no longer needed. Instead we would need a `isHidden` and `isAlias` or similar. Would also be nice to make the instantiation cheaper if `execute` is not needed (maybe bound statically with `Cnstancer` instead of the instantiation). Can you make an suggestion which does not take too much effort?

But let us better do this in a different PR, otherwise this gets never merged. So here you can keep it as internal command, simply rename it to `completion-helper` or `list-completions`.

markus2330 commented:
Btw. if all of you really want `super-ls` when can continue to use the name...

@e1528532 It is your decision

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+{
+	if (cl.arguments.size () != 1)
+	{
+		throw invalid_argument ("1 argument required");
```
Keep it consistent with others: `wrong number of arguments, 1 needed`.

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+	
+	function<bool(string)> pred = [] (string) { return true; };
+	bool isExistent = ks.lookup(root) || root.getBaseName().empty();
+	if (!isExistent) {
```
please run code formatter.

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+		string newKeyName = getParentKey(root).getFullName();
+		cout << "new name is " << newKeyName << endl;
+		root = Key (newKeyName, KEY_END);
+		cout << " building predicate " << root.isCascading() << " " << fullName << " " << fullName.find("/") << endl;
```
is this a verbose output? Please only if verbosity is set.

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+		root = Key (newKeyName, KEY_END);
+		cout << " building predicate " << root.isCascading() << " " << fullName << " " << fullName.find("/") << endl;
+		pred = [=] (string test) { 
+			return fullName.size() <= test.size() && std::equal(fullName.begin(), fullName.end(), test.begin() + (root.isCascading() ? test.find("/") : 0));
```
There is no requirement that lambdas fit in a single line.

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+		};
+	}
+	
+	cout << "cutting at " << root << endl;
```
should definitely be only in verbose

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+	
+	stack<Key> keyStack;
+	if (!(tmp = ks.next())) {
+		// return?
```
?

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+		return 0;
+	}
+	
+	int curDepth = 0;
```
function gets too long, please add more comments.

markus2330 replied:
maybe you can even create some reusable functions here?

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+	cout << "max is " << maxDepth << " and min is " << minDepth << endl;
+	
+	cout << "dumping all results" << endl;
+	for (auto it = hierarchy.begin(); it != hierarchy.end(); it++) {
```
there are even shorter for loops.

markus2330 commented on src/tools/kdb/superls.cpp:
```diff
@@ -0,0 +1,145 @@
+}
+
+void SuperLsCommand::increaseCount(map<Key, pair<int, int>> & hierarchy, Key key, function<int(int)> depthIncreaser) {
+	cout << "increasing for " << key << endl;
```
again, should be only in verbose.

markus2330 commented:
jenkins build fast please

markus2330 commented:
jenkins build please

e1528532 commented:
I think I failed to emphasize enough that its WIP and thus still contains some debugging output (i think most of the stuff i print right now wouldn't be wanted in a -v output and is really just for debugging). Also there is plenty of room for refactorings, and no standardized formatting yet (just using notepad++ right now), as i wanted to implement the functionality first and would have dealt with that afterwards.

However, it looks like overall you are more happy with my code than i currently am, and that using small functional abstractions is ok :) I had the feeling that larger adjustments are necessary.

markus2330 commented:
> I think I failed to emphasize enough that its WIP

There is a "work in progress" label just for that. But its better to give more feedback early than to have too many review rounds.

> and no standardized formatting yet (just using notepad++ right now), as i wanted to implement the functionality first and would have dealt with that afterwards.

Do not worry about editor or formatting. Formatting is completely left to `scripts/reformat-source`. But for the next time please run the script before creating the PR, it makes it easier to read the code and the build jobs will not fail (they check if the formatter was used).

> However, it looks like overall you are more happy with my code than i currently am, and that using small functional abstractions is ok :) I had the feeling that larger adjustments are necessary.

Yes, thus early PRs are important! It is just a helper tool, some shortcuts are okay here.

sanssecours commented:
> @sanssecours any opinion?

If you want to merge this pull request immediately, then you do not need to wait for my review. I am currently sick 🤒. Therefore it might take me some time to come back to this.

As for the name: I do not like `super-ls` either. How about `complete`, which I think is both descriptive and short?

markus2330 commented:
> If you want to merge this pull request immediately

In general it is better if PRs can be merged fast (less rebasing and forgetting about what the PR was about). In this case merging would be helpful so that @reox could continue his work on the auto completion.

> How about complete, which I think is both descriptive and short?

Sounds nice! In #265 we basically came to the conclusion that the names should contain what to do and where it should be done. E.g., not `list` but `list-plugins`. But in this case it directly works on keys, so `complete` actually would be fine. If we add completion for plugins, it would be `complete-plugins`.

Thanks for the input and get well soon!

e1528532 commented:
Ok I think I'm getting pretty close to finishing this.

The most important change i've made is that i decided to call the tool simply "complete" as sansseccours suggested and markus approved, because its our general completion tool. I agree that this name is better, as if I'd look at the list of commands kdb offers, I would have no idea what super-ls does at a first sight compared to ls, but complete already gives a good indication about its purpose.

I have also changed the interface now to my proposal:
-v to display verbose output (main difference: shows you the min/max output depths it will actually use and shows you if the path you want to complete does not exist or what else it uses as the reference path used for determining the depth)
-d to display debugging output (shows more details how it processed/expands keys and will give you the whole analyzation result prior to filtering it by name/depth
--min-depth to specifiy the minimum output depth, exclusive. 0 by default.
--max-depth to specify the maximum output depth, inclusive, 1 by default as i think when using a complete command you'd usually just want to see suggestions for the current level. -1 to show all levels.
-0 to have nulls as whitespace (stolen from kdb ls)

Some feedback would be appreciated if the depth parameters and defaults make sense.

The last big addition is that the command will also complete to mountpoints now, even if there are no keys in the mountpoint yet as requested by markus.

Furthermore i have refactored the code to separate the logic in (hopefully) meaningful functions and reviewed the output, displaying some debugging and verbose things only when the corresponding parameters are set. I also used our code formatter and tried to keep all your suggestions in mind.

I have a question about the getopt stuff though, i provided short options for my minimum and maximum depth (m for minimum and M for maximum), but i failed to use them. It works fine when i specify the options in the long format, like kdb complete -vd --min-depth=2 --max-depth=5 user/ , but fails when i try it in the short format like kdb complete -vd -m 2 -M 5 user/, -m2 or -m=2 doesn't work either. Is it not possible to have arguments for short options, and if so, can i hide the short option somehow?

Also, do we have some code linter doing some code analysis and monitoring things like unfreed references or complexity violations or similar I could/should use?
And do we maintain unit tests for the kdb tools? There are some tests ins libelektra/tests but they seem to be related to the main elektra library and not to any tools.

reox commented:
For min/max I would use the defaults of `find`: min 0 and max inf. 
I think the default should be to show everything, where other tools that might only need some subpart, can explicitly ask for it.

Am 21. Jänner 2017 17:07:34 MEZ schrieb e1528532 <notifications@github.com>:
>Ok I think I'm getting pretty close to finishing this.
>
>The most important change i've made is that i decided to call the tool
>simply "complete" as sansseccours suggested and markus approved,
>because its our general completion tool. I agree that this name is
>better, as if I'd look at the list of commands kdb offers, I would have
>no idea what super-ls does at a first sight compared to ls, but
>complete already gives a good indication about its purpose.
>
>I have also changed the interface now to my proposal:
>-v to display verbose output (main difference: shows you the min/max
>output depths it will actually use and shows you if the path you want
>to complete does not exist or what else it uses as the reference path
>used for determining the depth)
>-d to display debugging output (shows more details how it
>processed/expands keys and will give you the whole analyzation result
>prior to filtering it by name/depth
>--min-depth to specifiy the minimum output depth, exclusive. 0 by
>default.
>--max-depth to specify the maximum output depth, inclusive, 1 by
>default as i think when using a complete command you'd usually just
>want to see suggestions for the current level. -1 to show all levels.
>-0 to have nulls as whitespace (stolen from kdb ls)
>
>Some feedback would be appreciated if the depth parameters and defaults
>make sense.
>
>The last big addition is that the command will also complete to
>mountpoints now, even if there are no keys in the mountpoint yet as
>requested by markus.
>
>Furthermore i have refactored the code to separate the logic in
>(hopefully) meaningful functions and reviewed the output, displaying
>some debugging and verbose things only when the corresponding
>parameters are set. I also used our code formatter and tried to keep
>all your suggestions in mind.
>
>I have a question about the getopt stuff though, i provided short
>options for my minimum and maximum depth (m for minimum and M for
>maximum), but i failed to use them. It works fine when i specify the
>options in the long format, like kdb complete -vd --min-depth=2
>--max-depth=5 user/ , but fails when i try it in the short format like
>kdb complete -vd -m 2 -M 5 user/, -m2 or -m=2 doesn't work either. Is
>it not possible to have arguments for short options, and if so, can i
>hide the short option somehow?
>
>Also, do we have some code linter doing some code analysis and
>monitoring things like unfreed references or complexity violations or
>similar I could/should use?
>And do we maintain unit tests for the kdb tools? There are some tests
>ins libelektra/tests but they seem to be related to the main elektra
>library and not to any tools.
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ElektraInitiative/libelektra/pull/1281#issuecomment-274270489

markus2330 commented on README.md:
```diff
@@ -13,9 +13,9 @@ applications' configurations, leveraging easy application integration.
-- [build server](http://build.libelektra.org:8080/)
-- [tutorials](http://git.libelektra.org/blob/master/doc/tutorials/)
-- [API docu](http://doc.libelektra.org/api/latest/html/)
+- [build server](https://build.libelektra.org/)
```
It seems like you have severe rebasing/merging issues here!

e1528532 replied:
Looks like it. I've never used merge for this branch, and tried rebasing by using "git rebase ElektraInitiative/master" while i'm in this branch, but that somehow lead to this merge commit b970a1f ... isn't this the way to rebase things? It hasn't told me about any conflicts and hasn't asked me to resolve anything either so i wonder where this commit comes from.

markus2330 replied:
`git pull --rebase` should do the right thing. Now its the easiest to do a `cherry-pick` of the correct commits in a new branch.

e1528532 replied:
i slowly suspect that the github desktop app did all kinds of messy things. I've switched to sourcetree now and it think its much better, gives you a much better indication of what you are about to do and its very easy to rebase or work with branches. I just created a new branch based on the last good commit of super-ls, so no cherry picking needed. Will open a new pull request after dealing with the last few things here.

sanssecours replied:
> I just created a new branch based on the last good commit of super-ls, so no cherry picking needed.

Thank you for providing the new branch. Currently the code in `kdb-complete` does not compile on my machine. It works after I remove the files`superls.cpp` and `superls.hpp`, and replace 

```cpp
#include <superls.hpp>
```

in `complete.cpp` with

```cpp
#include <complete.hpp>
```

.

e1528532 replied:
ah yeah i've already fixed that but haven't comitted sorry, now it should be in the branch

markus2330 commented:
I would avoid having min/max for now. Is it really needed? It would be more useful for the actual `ls` (@tom-wa requested it for `ls` already).

> Also, do we have some code linter doing some code analysis and monitoring things like unfreed references or complexity violations or similar I could/should use?

See doc/TESTING.md. For memory issues we use valgrind and ASAN, complexity see #1248

> And do we maintain unit tests for the kdb tools?
> There are some tests ins libelektra/tests but they seem to be related to the main elektra library and not to any tools.

We have tests for libtools (the more complex stuff of what the tools do is implemented there, if the complete stuff gets too complicated we might even move it there) and we have the shellrecorder.

Btw. man page for the command is missing. (doc/help)

e1528532 commented:
@markus2330 IMHO having min and maxdepth is not a bad thing, as the completion tool is primarily intended for internal use / shell completion and things like that, so those two parameters give you some flexibility in how you can use it. 

manpage is added

valgrind seems unsupported on osx, running it with ASAN enabled gives me the following output:
```
/Users/admin/git/libelektra/src/libs/elektra/keyset.c:914:11: runtime error: unsigned integer overflow: 19 + 18446744073709551597 cannot be represented in type 'unsigned long'
SUMMARY: AddressSanitizer: undefined-behavior /Users/admin/git/libelektra/src/libs/elektra/keyset.c:914:11 in 
```

... That doesn't seem to be related to my code right?

And I've also checked the complexity, the analyze method has 13, but 3 are caused by output switches (if cl.verbose etc), so i'd say thats okay, the same for printResults (12, where 2 are caused by output switches)

so the last big thing to do is to check out how to create tests for it

markus2330 commented:
> IMHO having min and maxdepth is not a bad thing, as the completion tool is primarily intended for internal use / shell completion and things like that, so those two parameters give you some flexibility in how you can use it.

Flexibility is not necessarily a good thing if nobody needs it. But you can leave it for now if we need the same arguments for `ls` anyway.

> manpage is added

How is it called? I cannot find it.

> valgrind seems unsupported on osx

iirc people reported that it worked on some versions of osx. It is the best tool, so you might want to consider to have some virtual box/chroot/..

> running it with ASAN enabled gives me the following output:

Please also fix other ASAN issues if possible. You are even assigned for #788 ;)

> And I've also checked the complexity, the analyze method has 13, but 3 are caused by output switches (if cl.verbose etc), so i'd say thats okay, the same for printResults (12, where 2 are caused by output switches)

Please add this information as source code comment.

> so the last big thing to do is to check out how to create tests for it

If you already tested it manually I would prefer to get it merged quickly and get your concentration back to more important things. Do not get me wrong, auto-completion is important, but I am afraid this PR already does way too much (e.g. min/maxdepth).

e1528532 commented:
Ok so i'll leave the parameters for now.

manpage is called 'kdb-complete' (like the one for kdb ls is called kdb-ls), just tried it out it should work.

I'll fix the ASAN issues in a different pull request, as they don't seem to be directly related to this.

I added the mc cabe info to the source code for the two exceeding methods, the others are fine

And yeah i already did quite a lot of manual testing during development, so i've opened a new PR #1294 as in this branch there are these merge issues, and in the other not. Guess its ready to be merged then.
